### DISCUSSION TITLE
[MongoDB] Referential Action "NoAction" doesn't do what expected (actually does actions)

### DISCUSSION THREAD

**Discussion by ItzSiL3Nce** [prisma/prisma#20544] [Q&A] [answered]
### Bug description

I'm using the MongoDB datasource.

I've set the `NoAction` relation mode to both `OnUpdate` and `OnDelete` to every relation of my model(s).
Whenever I try to delete a document from the database, it starts checking and aggregating data from every other collection (from the relations) and ends up timing-out cause of the vast data I've got. 
I've seen this by enabling the `query` log.

Shouldn't the `NoAction` relation mode just let me delete the document without worrying of any "pending" relations?

### Expected behavior

Prisma should not do anything else other than deleting/updating the documents when `NoAction` is used.

### Prisma information

```prisma
generator client {
  provider      = "prisma-client-js"
  output        = "./prisma/prisma-client"
  binaryTargets = ["native"]
}

datasource db {
  provider = "mongodb"
  url      = env("DATABASE_URL")
}

model MockRecord {
    id            String         @id @default(auto()) @map("_id") @db.ObjectId
    mockData      String
    mockRelations MockRelation[]
}

model MockRelation {
    id           String     @id @default(auto()) @map("_id") @db.ObjectId
    mockRecord   MockRecord @relation(fields: [mockRecordId], references: [id], onDelete: NoAction, onUpdate: NoAction)
    mockRecordId String     @db.ObjectId
}
```

```ts
import { PrismaClient } from './prisma/prisma-client'

const prisma = new PrismaClient({
  log: ['query']
})

const newRecord = await prisma.mockRecord.create({ 
  data: { mockData: "test" } 
})

// Upon executing the delete, prisma will start aggregating things from the collections it has references to/from.
// No matter the "NoAction" on the "onDelete".
await prisma.mockRecord.delete({
 where: {
  id: newRecord.id
 }
})
```

### Environment & setup

- OS: Ubuntu
- Database: MongoDB
- Node.js version: 18.13.0

### Prisma Version

```
prisma                  : 5.1.0
@prisma/client          : 5.1.0
Current platform        : debian-openssl-3.0.x
Query Engine (Node-API) : libquery-engine a9b7003df90aa623086e4d6f4e43c72468e6339b (at ../../node_modules/.pnpm/@prisma+engines@5.1.0/node_modules/@prisma/engines/libquery_engine-debian-openssl-3.0.x.so.node)
Schema Engine           : schema-engine-cli a9b7003df90aa623086e4d6f4e43c72468e6339b (at ../../node_modules/.pnpm/@prisma+engines@5.1.0/node_modules/@prisma/engines/schema-engine-debian-openssl-3.0.x)
Schema Wasm             : @prisma/prisma-schema-wasm 5.1.0-28.a9b7003df90aa623086e4d6f4e43c72468e6339b
Default Engines Hash    : a9b7003df90aa623086e4d6f4e43c72468e6339b
Studio                  : 0.492.0
```

**Comment by janpio** [comment, marked as answer]
The naming of `NoAction`, which comes from SQLs `NO ACTION` is unfortunately not the clearest. Here is the section in our docs: https://www.prisma.io/docs/concepts/components/prisma-schema/relations/referential-actions#noaction It means that "no action" is taken (vs. "cascade" or "set null"), and as that would lead to invalid data (some relation field would point to an entry that does not exist), the action is **prevented**.

There is currently no "do nothing and allow invalid data to be created" option in Prisma's referential actions.

**Reply by ItzSiL3Nce** [reply]
Oh, thanks for explaining my misunderstanding, now it makes sense.
Is it possible to convert this in a Feature Request? Do you think what I'm asking for is doable?
My company has switched recently to Prisma and it'd be very handful for us, and probably others who want to use MongoDB with a "unsafer" relation model, like it's supposed to be.

I'll probably make a pull request anyway, because I'm going to make my own version with this feature.

**Reply by janpio** [reply]
This is definitely a valid feature request. Unfortunately a pretty challenging one for us to implement, as this goes against how Prisma _wants_ things to work - if there is a "relation" pointing to a value that does not exist, that makes Prisma pretty unhappy and it often crashes (This is because in SQL databases with foreign keys, this is practically impossible). So part of that feature would be to find all these cases in our code and return a better error message.

There might already be feature requests related to this, please search a bit before you open an issue if you can just comment on an existing one and provide your use case and context. If you can't find one, go ahead though and create one. I totally understand why this is useful when using MongoDB.

**Reply by Jolg42** [reply]
I think this existing issue captures the feature request here: https://github.com/prisma/prisma/issues/15759

**Comment by ItzSiL3Nce** [comment]
I've made a pull request for this feature: https://github.com/prisma/prisma-engines/pull/4116

**Comment by bennet-esyoil** [comment]
Hello,
very happy so see this here. We have some issues deleting single entries that have multiple thousand relations (everything between a few single ones up to 250k). Those queries would almost always timeout. The `NoAction` is in-fact a bit misleading and the implemented solution is imo the best way to tackle this problem.
Otherwise one would need to fallback to the `$runCommandRaw` which is reaaally ugly.. I'd be more than happy to see the contribution merged, otherwise there is nothing really developers can do about those big performance hits, other than to accept them.

I am also a bit confused that there are not more issues related to this but it also took me some time to find this one.
